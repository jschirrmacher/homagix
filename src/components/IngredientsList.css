.IngredientsList li {
  padding-right: 30px;
}

.IngredientsList li button.inline {
  right: -30px;
}

.IngredientsList li:last-of-type {
  padding-right: 15px;
}

.ItemGroup  {
  float: right;
  font-size: 15px;
  border: none;
  background: transparent;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  outline: none;
  padding: 2px 10px;
  border-radius: 3px;
  -ms-text-align-last: center;
  text-align-last: center;
  margin-right: -25px;
}

#additionalItem input {
  display: inline-block;
  font-size: 16px;
  width: calc(100% - 25px);
  padding: 5px;
  border: 1px solid lightgrey;
}

#additionalItem:after {
  content: '+';
  display: inline-block;
  position: absolute;
  right: 3px;
  font-size: 28px;
  color: #009900;
}

.modified {
  color: white;
  transition: all .5s;
}

@media screen and (max-width: 640px) {
  .ItemGroup {
    padding: 5px 0;
    font-size: 11px;
  }
}

@media screen {
  .unit {
    display: inline-block;
    width: 50px;
    margin: 0 5px;
  }

  .amount {
    display: inline-block;
    width: 50px;
    text-align: right;
  }
}

.color-fruit {
  background-color: darkgreen;
  color: yellow;
}

.color-breakfast {
  background-color: yellow;
  color: saddlebrown;
}

.color-meat {
  background-color: darkred;
  color: yellow;
}

.color-cooled {
  background-color: cornflowerblue;
  color: white;
}

.color-tinned {
  background-color: lightgrey;
  color: black;
}

.color-drinks {
  background-color: yellowgreen;
  color: yellow;
}

.color-frozen {
  background-color: blue;
  color: white;
}
