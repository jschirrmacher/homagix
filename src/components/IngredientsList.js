import React, { Component } from 'react'
import { connect } from 'react-redux'
import { setItemGroup } from '../actions/proposalsActions'
import { bindActionCreators } from 'redux'
import './IngredientsList.css'

let index = 0

const itemGroups = {
  fruit: { order: 1, title: 'Obst & Gemüse' },
  breakfast: { order: 2, title: 'Brot & Frühstück' },
  meat: { order: 3, title: 'Fleisch' },
  cooled: { order: 4, title: 'Frische & Kühlung' },
  tinned: { order: 5, title: 'Nahrungsmittel' },
  drinks: { order: 6, title: 'Getränke' },
  frozen: { order: 7, title: 'Tiefgekühlt' },
  other: { order: 8, title: 'Sonstiges' }
}

const knownUnits = ['Stk', 'Pkg', 'Glas', 'Zehen', 'Würfel', 'Dose', 'Kopf', 'Bund', 'g', 'kg', 'L', 'ml', 'cm']

function guessUnit(text) {
  return knownUnits.find(unit => text.match(new RegExp('\\b' + unit + '\\.?\\b', 'i')))
}

function classNames(props) {
  return Object.keys(props).filter(name => !!props[name]).join(' ')
}

class IngredientsList extends Component {
  constructor(props) {
    super(props)
    this.state = { inhibit: {}, additions: [], modified: null, inputChanged: false }
  }

  removeIngredient(item, elem) {
    if (item.id < 0) {
      elem.classList.add('fading')
      setTimeout(() => this.setState(state => {
        state.additions = state.additions.filter(el => el.id !== item.id)
        return state
      }), 1000)
    } else {
      this.setState(state => {
        state.inhibit[item.id] = !state.inhibit[item.id]
        return state
      })
    }
  }

  add(item, list) {
    const existing = list.find(entry => entry.name === item.name)
    if (existing && existing.unit === item.unit) {
      existing.amount += item.amount
    } else {
      item.id = --index
      list.push(item)
    }
  }

  addIngredient(elem) {
    if (elem.value && this.state.inputChanged) {
      this.setState(state => {
        const amount = parseFloat(elem.value.replace(/(\d),(\d)/, '$1.$2')) || 1
        const rest = elem.value.replace(new RegExp(('' + amount).replace(/\./, '[.,]')), '')
        const unit = guessUnit(rest) || 'Stk'
        const name = rest.replace(new RegExp(unit + '\\.?', 'i'), '').trim()
        this.add({ amount, unit, name }, state.additions)
        state.modified = name
        setTimeout(() => this.setState({ modified: null }), 500)
        elem.value = ''
        return state
      })
    }
  }

  handleKeyDown(event) {
    if (event.keyCode === 13) {
      this.addIngredient(event.target)
    }
    this.setState({ inputChanged: true })
  }

  selectIngredient(el, text) {
    if (['SELECT', 'BUTTON'].includes(el.tagName)) {
      return
    }
    const newItem = document.querySelector('#additionalItem input')
    newItem.value = text
    newItem.setSelectionRange(0, 0)
    newItem.focus()
    this.setState({ inputChanged: false })
  }

  getIngredients() {
    const ingredients = (this.props.ingredients || []).map(ingredient => ({ ...ingredient }))
    this.state.additions.forEach(add => this.add(add, ingredients))
    ingredients.sort((a, b) => {
      const comp = itemGroups[a.group || 'other'].order - itemGroups[b.group || 'other'].order
      return comp || a.name.localeCompare(b.name)
    })
    return ingredients
  }

  setItemGroup(item, group) {
    if (item.id > 0) {
      this.props.setItemGroup(item, group)
    } else {
      this.setState({ additions: this.state.additions.map(i => i.id === item.id ? Object.assign(i, { group }) : i) })
    }
  }

  render() {
    const items = this.getIngredients()
      .map(item => {
        const group = <select onChange={event => this.setItemGroup(item, event.target.value)}
          defaultValue={item.group} className={'ItemGroup color-' + item.group}>
          <option value={null}>-- bitte auswählen --</option>
          {Object.keys(itemGroups).map(group => (
            <option key={group} value={group}>{itemGroups[group].title}</option>
          ))}
        </select>

        const classes = classNames({ inhibited: this.state.inhibit[item.id], modified: this.state.modified === item.name })
        return <li key={item.id}
                   className={classes}
                   onClick={event => this.selectIngredient(event.target, item.unit + ' ' + item.name)}
               >
          <button className="delete inline" onClick={event => this.removeIngredient(item, event.target.parentNode)}>
            &times;
          </button>
          {group}
          {this.props.showAmount && <span className="amount">{item.amount}</span>}
          <span className="unit">{item.unit}</span>
          {item.name}
        </li>
      })

    return <ul className="IngredientsList">
      {items}
      {this.props.showNew && <li id="additionalItem">
        <input type="text"
          onKeyDown={event => this.handleKeyDown(event)}
          onBlur={event => this.addIngredient(event.target)}
        />
      </li>}
    </ul>
  }
}

function mapStateToProps() {
  return {}
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators({ setItemGroup }, dispatch)
}

export default connect(mapStateToProps, mapDispatchToProps)(IngredientsList)

