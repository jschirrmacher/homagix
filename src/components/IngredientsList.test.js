import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'
import renderer from 'react-test-renderer'
import {Provider} from 'react-redux'
import IngredientsList from './IngredientsList'
import 'should'

import configureStore from 'redux-mock-store'

const middlewares = []
const mockStore = configureStore(middlewares)

const initialState = {
  error: {},
  proposals: {ingredients: [{id: 4, amount: 5, unit: 'L', name: 'Milch'}]}
}

const store = mockStore(initialState)

describe('IngredientsList', () => {
  it('renders without crashing', () => {
    const renderer = new ShallowRenderer()
    renderer.render(<Provider store={store}><IngredientsList/></Provider>)
    const tree = renderer.getRenderOutput()
    expect(tree).toMatchSnapshot()
  })

  it('should render ingredients', () => {
    const tree = renderer.create(<Provider store={store}><IngredientsList/></Provider>)
    const li = tree.toJSON().children[0]
    const content = li.children.splice(1).join('')
    content.should.equal('5 L Milch')
  })

  it('has a button to remove proposed ingredients', () => {
    const tree = renderer.create(<Provider store={store}><IngredientsList/></Provider>)
    const li = tree.toJSON().children[0]
    li.children[0].type.should.equal('button')
    const button = li.children[0]
    button.props.className.should.equal('delete')
    button.children[0].should.equal('Ã—')
  })

  it('should mark items when the delete button is clicked', () => {
    const tree = renderer.create(<Provider store={store}><IngredientsList/></Provider>)
    const li = tree.toJSON().children[0]
    const button = li.children[0]
    button.props.onClick()
    tree.toJSON().children[0].props.className.should.equal('inhibited')
  })

  it('should remove the mark when the button is clicked again', () => {
    const tree = renderer.create(<Provider store={store}><IngredientsList/></Provider>)
    const li = tree.toJSON().children[0]
    const button = li.children[0]
    button.props.onClick()
    button.props.onClick()
    should(tree.toJSON().children[0].props.className).be.undefined()
  })

  it('has an input field to add ingredients', () => {
    const tree = renderer.create(<Provider store={store}><IngredientsList/></Provider>)
    const input = tree.toJSON().children[1].children[0]
    input.type.should.equal('input')
    input.props.id.should.equal('additionalItem')
  })

  it('should add ingredients when enter is pressed', () => {
    const tree = renderer.create(<Provider store={store}><IngredientsList/></Provider>)
    const input = tree.toJSON().children[1].children[0]
    input.props.value = '2 Stk. Butter'
    input.props.onKeyDown({keyCode: 13, target: input.props})
    const li = tree.toJSON().children[1]
    const content = li.children.splice(1).join('').trim()
    content.should.equal('2 Stk. Butter')
  })

  it('should remove additional ingredients when delete button is pressed', () => {
    const tree = renderer.create(<Provider store={store}><IngredientsList/></Provider>)
    const input = tree.toJSON().children[1].children[0]
    input.props.value = '2 Stk. Butter'
    input.props.onKeyDown({keyCode: 13, target: input.props})
    const button = tree.toJSON().children[1].children[0]
    button.props.onClick()
    tree.toJSON().children[1].children[0].props.id.should.equal('additionalItem')
  })
})
