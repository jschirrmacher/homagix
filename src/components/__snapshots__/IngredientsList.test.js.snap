// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IngredientsList renders without crashing 1`] = `<Connect(IngredientsList) />`;
